File: app/api/models/Painting.py
from django.db import models
from django.utils.timezone import now

from .Creator import Creator
from .Depiction import Depiction
from .Genre import Genre
from .Location import Location
from .Material import Material
from .Movement import Movement


class Painting(models.Model):
    name = models.CharField(max_length=500)
    native_name = models.CharField(max_length=500)
    wikidata_url = models.CharField(max_length=500, null=False)

    title = models.CharField(max_length=500)
    picture_url = models.CharField(max_length=500)
    owned_by = models.CharField(max_length=500)
    inception_at = models.DateTimeField()
    width = models.PositiveIntegerField()
    height = models.PositiveIntegerField()
    described_at = models.CharField(max_length=500)

    creators = models.ManyToManyField(Creator)
    movements = models.ManyToManyField(Movement)
    genres = models.ManyToManyField(Genre)
    depicts = models.ManyToManyField(Depiction)
    materials = models.ManyToManyField(Material)
    locations = models.ManyToManyField(Location, related_name='actual_location')
    location_of_creations = models.ManyToManyField(Location, related_name='creation_location')

    created_at = models.DateTimeField(default=now)
    updated_at = models.DateTimeField(default=now)

File: app/Models/Painting.php
<?php

namespace App\Models;

use App\Traits\FetchingRegisterable;
use App\Traits\FindableByUniqueKey;
use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

class Painting extends Model
{
  use HasFactory;
  use FindableByUniqueKey;
  use FetchingRegisterable;

  protected static function boot()
  {
    parent::boot();
  }

  protected $guarded = ['id'];
  protected $casts = [
    'aliases' => 'array',
    'picture_url' => 'array',
    'title' => 'array',
    'owned_by' => 'array',
    'inception_at' => 'array',
    'width' => 'array',
    'height' => 'array',
    'described_at' => 'array',
  ];

  public static function mapping(string $sourceName): ?array
  {
    $mappings = [
      'wikidata' => [
        'oneToOne' => [
          'label' => 'name',
          'aliases' => 'aliases',
          'description' => 'description',
          'id' => 'wikidata_id'
        ],
        'manyToOne' => [
          'properties.P18.values.0.label'  => 'picture_url',
          'properties.P1476.values.0.label' => 'title',
          'properties.P127.values.0.label' => 'owned_by',
          'properties.P571.values.0.label' => 'inception_at',
          'properties.P2049.values.0.label' => 'width',
          'properties.P2048.values.0.label' => 'height',
          'properties.P973.values.0.label' => 'described_at'
        ],
        'relations' => [
          'properties.P170.values.0.id' => Artist::class,
          'properties.P180.values.0.id' => Depiction::class,
          'properties.P136.values.0.id' => Genre::class,
          'properties.P276.values.0.id' => Location::class,
          'properties.P186.values.0.id' => Material::class,
          'properties.P135.values.0.id' => Movement::class,
        ]
      ]
    ];

    return $mappings[$sourceName] ?? null;
  }

  public function artists()
  {
    return $this->belongsToMany(Artist::class);
  }

  public function depictions()
  {
    return $this->belongsToMany(Depiction::class);
  }

  public function genres()
  {
    return $this->belongsToMany(Genre::class);
  }

  public function locations()
  {
    return $this->belongsToMany(Location::class);
  }

  public function materials()
  {
    return $this->belongsToMany(Material::class);
  }

  public function movements()
  {
    return $this->belongsToMany(Movement::class);
  }
}
